### PR TITLE
[search] Support for categorial search in the geocoder.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

**Comment by mpimenov** [comment]
@ygorshenin @tatiana-kondakova PTAL

File: search/categories_cache.cpp [hunk 1]
```diff
@@ -31,14 +31,35 @@ CBV CategoriesCache::Get(MwmContext const & context)
   return cbv;
 }
 
+CBV CategoriesCache::GetFuzzy(MwmContext const & context)
+{
+  if (!context.m_handle.IsAlive() || !context.m_value.HasSearchIndex())
+    return CBV();
+
+  auto id = context.m_handle.GetId();
```
Comment (tatiana-kondakova) [inline search/categories_cache.cpp]: const?

File: search/categories_cache.cpp [hunk 2]
```diff
@@ -31,14 +31,35 @@ CBV CategoriesCache::Get(MwmContext const & context)
...
+  auto const it = m_cacheFuzzy.find(id);
+  if (it != m_cacheFuzzy.cend())
+    return it->second;
+
+  auto cbv = LoadFuzzy(context);
```
Comment (tatiana-kondakova) [inline search/categories_cache.cpp]: const?

File: search/categories_cache.cpp [hunk 3]
```diff
@@ -48,6 +69,23 @@ CBV CategoriesCache::Load(MwmContext const & context)
   return CBV(retrieval.RetrieveAddressFeatures(request));
 }
 
+CBV CategoriesCache::LoadFuzzy(MwmContext const & context)
```
Comment (tatiana-kondakova) [inline search/categories_cache.cpp]: кажется этот метод не модифицирует поля класса и может быть константным

File: search/categories_cache.cpp [hunk 4]
```diff
@@ -31,14 +31,35 @@ CBV CategoriesCache::Get(MwmContext const & context)
   return cbv;
 }
 
+CBV CategoriesCache::GetFuzzy(MwmContext const & context)
+{
+  if (!context.m_handle.IsAlive() || !context.m_value.HasSearchIndex())
```
Comment (ygorshenin) [inline search/categories_cache.cpp]: We have an agreement that context is always valid here, no no need in these m_handle.IsAlive(). Also, no need in check that context has search index.

Comment (mpimenov) [reply search/categories_cache.cpp]: I'm not sure how it is bad to be on the safer side.
That's old code and I don't see the agreement in its docs. Do you think I shall write it and replace the "if" with asserts?

File: search/categories_cache.cpp (hunk 1 above)
Comment (mpimenov) [reply search/categories_cache.cpp]: done

File: search/categories_cache.cpp (hunk 2 above)
Comment (mpimenov) [reply search/categories_cache.cpp]: done

File: search/categories_cache.cpp (hunk 3 above)
Comment (mpimenov) [reply search/categories_cache.cpp]: done

File: search/categories_cache.cpp (hunk 4 above)
Comment (mpimenov) [reply search/categories_cache.cpp]: done

**Comment by mpimenov** [comment]
@ygorshenin @tatiana-kondakova PTAL

File: search/categories_cache.hpp [hunk 5]
```diff
@@ -23,14 +24,21 @@ class CategoriesCache
     source.ForEachType([this](uint32_t type) { m_categories.Add(type); });
   }
 
+  CategoriesCache(set<uint32_t> types, my::Cancellable const & cancellable)
```
Comment (ygorshenin) [inline search/categories_cache.hpp]: Const ref for types, please.

File: search/geocoder.cpp [hunk 6]
```diff
@@ -434,6 +440,32 @@ void Geocoder::GoInViewport()
...
+  m_matchersCache.clear();
+  m_streetsCache.Clear();
+  m_hotelsCache.Clear();
+  m_hotelsFilter.ClearCaches();
+  m_postcodes.Clear();
+}
+
+void Geocoder::SetParamsForCategorialSearch(Params const & params)
+{
+  m_params = params;
```
Comment (ygorshenin) [inline search/geocoder.cpp]: This code duplicates part of code from SetParams, could you please get rid of this duplication?

File: search/geocoder.cpp [hunk 7]
```diff
@@ -565,14 +591,22 @@ void Geocoder::InitBaseContext(BaseContext & ctx)
...
-      ctx.m_features[i] = retrieval.RetrieveAddressFeatures(m_prefixTokenRequest);
+    if (m_params.IsCategorialRequest())
+    {
+      // Implementation-wise, the simplest way to match a feature by
+      // its category bypassing the matching by name is by using a CategoriesCache.
+      CategoriesCache cache(m_params.m_preferredTypes, m_cancellable);
+      ctx.m_features[i] = cache.Get(*m_context);
+    }
+    else if (m_params.IsPrefixToken(i))
+      ctx.m_features[i] = retrieval.RetrieveAddressFeaturesFuzzy(m_prefixTokenRequest);
```
Comment (ygorshenin) [inline search/geocoder.cpp]: Curly braces here, please.

File: search/geocoder.cpp [hunk 8]
```diff
@@ -735,6 +776,24 @@ void Geocoder::ForEachCountry(vector<shared_ptr<MwmInfo>> const & infos, TFn &&
...
+  auto emit = [&](uint64_t bit) {
+    auto const featureId = base::asserted_cast<uint32_t>(bit);
+    Model::Type type;
+    if (!GetTypeInGeocoding(ctx, featureId, type))
+      return;
+
+    EmitResult(ctx, m_context->GetId(), featureId, type, TokenRange(0, 1), nullptr /* geoParts */);
+  };
+
+  // By now there's only one token and zero prefix tokens.
```
Comment (ygorshenin) [inline search/geocoder.cpp]: Where the intersection with pivot is done?

File: search/query_params.hpp [hunk 9]
```diff
@@ -115,6 +115,9 @@ class QueryParams
   inline Langs const & GetLangs() const { return m_langs; }
   inline bool LangExists(int8_t lang) const { return m_langs.Contains(lang); }
 
+  inline void SetCategorialRequest(bool rhs) { m_isCategorialRequest = rhs; }
```
Comment (ygorshenin) [inline search/query_params.hpp]: s/rhs/categorical/

File: search/query_params.hpp [hunk 10]
```diff
@@ -123,6 +126,7 @@ class QueryParams
   vector<Token> m_tokens;
   Token m_prefixToken;
   bool m_hasPrefix = false;
+  bool m_isCategorialRequest = false;
```
Comment (ygorshenin) [inline search/query_params.hpp]: I propose you to rename this to isCategorialRequest.

File: search/retrieval.hpp [hunk 11]
```diff
@@ -38,9 +38,15 @@ class Retrieval
   // Following functions retrieve from the search index corresponding to
   // |value| all features matching to |request|.
   unique_ptr<coding::CompressedBitVector> RetrieveAddressFeatures(
-      SearchTrieRequest<strings::LevenshteinDFA> const & request);
+      SearchTrieRequest<strings::UniStringDFA> const & request);
 
   unique_ptr<coding::CompressedBitVector> RetrieveAddressFeatures(
+      SearchTrieRequest<strings::PrefixDFAModifier<strings::UniStringDFA>> const & request);
+
+  unique_ptr<coding::CompressedBitVector> RetrieveAddressFeaturesFuzzy(
```
Comment (ygorshenin) [inline search/retrieval.hpp]: Could you please revert these changes?

File: search/utils.hpp [hunk 12]
```diff
@@ -118,4 +120,31 @@ void ForEachCategoryTypeFuzzy(StringSliceBase const & slice, TLocales const & lo
...
+bool IsCategorialRequest(QuerySliceOnRawStrings<T> const & slice, Locales const & locales,
+                         CategoriesHolder const & catHolder)
+{
+  if (slice.Size() != 1 || slice.HasPrefixToken())
+    return false;
+
+  bool found = false;
+  auto token = slice.Get(0);
+  catHolder.ForEachName([&](CategoriesHolder::Category::Name const & categorySynonym) {
+    if (std::find(locales.begin(), locales.end(), categorySynonym.m_locale) == locales.end())
```
Comment (ygorshenin) [inline search/utils.hpp]: nit: you can use SmallSet for locales.

File: search/categories_cache.hpp (hunk 5 above)
Comment (mpimenov) [reply search/categories_cache.hpp]: done

File: search/geocoder.cpp (hunk 6 above)
Comment (mpimenov) [reply search/geocoder.cpp]: I couldn't find a neat way to do it. If you insist, I can factor out those 4 lines to a helper method.
Note that the log message must be duplicated even if it is the same (the log prefix contains the call site).

File: search/geocoder.cpp (hunk 7 above)
Comment (mpimenov) [reply search/geocoder.cpp]: done

File: search/geocoder.cpp (hunk 8 above)
Comment (mpimenov) [reply search/geocoder.cpp]: https://github.com/mapsme/omim/pull/7213/files/2846b625ed4d5644b76f8ece3a3d30226f978db7#diff-7e907361d64c0be8c2f5b2053636ef1eR542

File: search/query_params.hpp (hunk 9 above)
Comment (mpimenov) [reply search/query_params.hpp]: Yup, I hesitated here. Changed to isCategorial

File: search/query_params.hpp (hunk 10 above)
Comment (mpimenov) [reply search/query_params.hpp]: What do you mean? Remove the "m_"? It's against the style guide.

File: search/retrieval.hpp (hunk 11 above)
Comment (mpimenov) [reply search/retrieval.hpp]: I can, but why?

File: search/utils.hpp (hunk 12 above)
Comment (mpimenov) [reply search/utils.hpp]: I like the idea, but let's do it in a separate PR, and for all places that use Locales at once.

File: search/retrieval.hpp (hunk 11 above)
Comment (mpimenov) [reply search/retrieval.hpp]: Removed "Fuzzy" from the method names.

**Comment by mpimenov** [comment]
@ygorshenin @tatiana-kondakova PTAL

File: search/geocoder.cpp (hunk 8 above)
Comment (mpimenov) [reply search/geocoder.cpp]: Although, additional intersection in the non-viewport case won't hurt. Working on it.

Comment (mpimenov) [reply search/geocoder.cpp]: done

File: search/geocoder.cpp [hunk 13]
```diff
@@ -524,11 +556,18 @@ void Geocoder::GoImpl(vector<shared_ptr<MwmInfo>> & infos, bool inViewport)
                        m_cities = citiesFromWorld;
                      });
 
+      if (m_params.IsCategorialRequest())
+      {
+        bool const aroundPivot = index < numIntersectingMaps;
```
Comment (ygorshenin) [inline search/geocoder.cpp]: Please, move this out of the if and reuse in both branches.

Comment (mpimenov) [reply search/geocoder.cpp]: done
